src/etcd/client.py

```diff
@@ -572,11 +575,29 @@ def api_execute(self, path, method, params=None, timeout=None):
                         fields=params,
                         timeout=timeout,
                         encode_multipart=False,
-                        redirect=self.allow_redirect)
+                        redirect=self.allow_redirect,
+                        preload_content=False)
                 else:
                     raise etcd.EtcdException(
                         'HTTP method {} not supported'.format(method))
 
+                # If requested, check the cluster ID hasn't changed under us.
+                # We need preload_content == False above to ensure we can read
+                # the headers here before waiting for the content of a watch
+                # below.
+                cluster_id = response.getheader("x-etcd-cluster-id")
+                if (check_cluster_uuid and
+                        self._cluster_uuid and
+                        (cluster_id != self._cluster_uuid)):
+                    raise etcd.EtcdClusterIdChanged(
+                        'The UUID of the cluster changed from {} to '
+                        '{}.'.format(self._cluster_uuid, cluster_id))
+                elif not self._cluster_uuid:
+                    self._cluster_uuid = cluster_id
```

Comment:
I don't like having all this logic inside the `try`: can we pull it out, either into an `else` block or after the `try...except`?